core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -933,17 +933,26 @@ object ZSTM {
    * Atomically performs a batch of operations in a single transaction.
    */
   def atomically[R, E, A](stm: ZSTM[R, E, A]): ZIO[R, E, A] =
-    ZIO.accessM[R] { r =>
-      ZIO.effectSuspendTotalWith { (platform, fiberId) =>
-        tryCommit(platform, fiberId, stm, r) match {
-          case TryCommit.Done(io) => io // TODO: Interruptible in Suspend
-          case TryCommit.Suspend(journal) =>
-            val txnId     = makeTxnId()
-            val done      = new AtomicBoolean(false)
-            val interrupt = UIO(Sync(done)(done.set(true)))
-            val async     = ZIO.effectAsync(tryCommitAsync(journal, platform, fiberId, stm, txnId, done, r))
-
-            async ensuring interrupt
+    ZIO.uninterruptibleMask { restore =>
+      ZIO.accessM[R] { r =>
+        ZIO.effectSuspendTotalWith { (platform, fiberId) =>
+          val state = new AtomicReference[State[E, A]](State.Running)
+          tryCommit(platform, fiberId, stm, state, r) match {
+            case TryCommit.Done(io) => io // TODO: Interruptible in Suspend
+            case TryCommit.Suspend(journal) =>
+              val txnId     = makeTxnId()
+              val interrupt = ZIO.effectTotal(state.compareAndSet(State.Running, State.Interrupted))
+              val async     = ZIO.effectAsync(tryCommitAsync(journal, platform, fiberId, stm, txnId, state, r))
+
+              restore(async).onInterrupt(interrupt).run.flatMap {
```

Comment:
`onInterrupt` is way more costly than `ensuring`, and `onInterrupt(...).run` ends up calling `run` two times. Overall, this section of code will have significant performance implications.

Reply:
I simplified this substantially to use a single fold.